### PR TITLE
Add autorecover-alarm action

### PR DESCRIPTION
**Pull request by jesselang**
Pairs nicely with the singleton filter.

### CODE REVIEW

File: tests/test_ec2.py [hunk 1]
```diff
@@ -904,6 +904,25 @@ def test_ec2_add_security_groups(self):
...
+                 {'type': 'singleton'}],
+             'actions': [
+                 {'type': 'autorecover-alarm'}]},
+            config={'region': 'us-west-1'},
+            session_factory=session_factory)
+
+        resources = p.run()
+
+        self.assertEqual(len(resources), 1)
+        self.assertEqual(resources[0]['InstanceId'], 'i-00fe7967fb7167c62')
```
Comment (kapilt) [inline tests/test_ec2.py]: this doesn't actually assert/verify the effect of the action which would be creating the named alarm "recover-{}'.format(i['InstanceId'])",  

as pseudo example of verifying
```
client = session_factory('cloudwatch')
try:
    result = client.describe_alarms(
         AlarmNames=['recover-{}'.format(resources[0]['InstanceId']])
    self.assertTrue(result.get('MetricAlarms')) 
except:
    self.fail("alarm not found")
```

Comment (jesselang) [inline tests/test_ec2.py]: Thanks for pointing that out. Will amend.

Comment (jesselang) [inline tests/test_ec2.py]: Should be g2g. Thanks!

**Comment by jesselang** [comment]
I realize these changes bled into the `singleton` filter a bit. I realized there was an error in the filter, so took the opportunity to correct and allow the action to use the same logic.

File: c7n/resources/ec2.py [hunk 2]
```diff
@@ -206,6 +206,34 @@ def get_resource_date(self, i):
...
+class ASGFilter(object):
+    """Filter instances by whether they are members of an ASG"""
+    asg_member = None
+
+    def filter_asg_membership(self, instances, desired=None):
+        if desired is None:
+            desired = self.asg_member
+
+        if self.asg_member is None:
+            # TODO raise exception
```
Comment (jesselang) [inline c7n/resources/ec2.py]: Looking for guidance on what to raise here. Or maybe set `asg_member = True` above?

Comment (kapilt) [inline c7n/resources/ec2.py]: its unclear that this filter should even exist. its already available as a value filter on the tag presence. where we have existing filters that provide the same capability there's a strong preference to use those vs adding in one-offs.

ie. the equivalent value filter
```
type: value
key: "tag:aws:autoscaling:groupName"
value: not-null
```

Comment (jesselang) [inline c7n/resources/ec2.py]: Is there a case where someone would want an `autorecover-alarm` on an ASG instance? I didn't think so, which is why I wanted to reuse the logic to exclude ASG instances in that action that I had used in the `singleton` filter. The `ASGFilter` class is there only to reuse the logic. I'm open to other ideas. Please advise.

Comment (kapilt) [inline c7n/resources/ec2.py]: its fine if the autorecover alarm inlines the filter (either composition or direct), i'm just cautious about including a separate filter for this, vs. just documenting the syntax for using the value type to filter correctly in the docs for the autorecover alarm action.

**Comment by jesselang** [comment]
Thanks for the review. I've utilized the `ValueFilter` to do the ASG filtering, and it's pretty obvious that was a better solution. Thanks for your help!

**Comment by jesselang** [comment]
Any more thoughts here @kapilt ? Thanks!

**Pull request merged**